// components-vitalii/restaurant-page/restaurant-page.js
import React, { Component, Fragment } from "react";
import "./restaurant-page.css";
import { withRouter } from "react-router-dom";
import Spinner from "components/spinner";
import RestaurantView from "./restaurant-view";
import Header from "components/shared-components/header";
import { getRestaurant } from "api";

const navItems = [
  { label: "Dishes", link: "/dishes" },
  { label: "Restaurants", link: "/restaurants" },
  { label: "Login", link: "/login" },
  { label: "Registration", link: "/registration" }
];

class RestaurantPage extends Component {
  state = {
    openDish: false,
    dishToShow: {}
  };
  componentDidMount() {
    console.log("props", this.props.match.params.id);
    getRestaurant(this.props.match.params.id).then(restaurant => {
      console.log("GETTED REST", restaurant);
      this.setState({ restaurant });
    });
  }
  onDishAddClick = dish => {
    this.setState({ dishToShow: dish });
    this.setState({ openDish: true });
  };

  render() {
    const { restaurant, openDish, dishToShow } = this.state;
    const content = restaurant ? (
      <RestaurantView
        data={{
          restaurant,
          openDish,
          dishToShow,
          onDishAddClick: this.onDishAddClick
        }}
      />
    ) : (
      <Spinner />
    );
    return (
      <Fragment>
        <Header navItems={navItems} />
        {content}
      </Fragment>
    );
  }
}
export default withRouter(RestaurantPage);
