// components/pages/schedule-items-page/schedule-item-dishes-list/schedule-item-dishes-list.js
import React from "react";
import PropTypes from "prop-types";
import Badge from "@material-ui/core/Badge";
import { withStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import Avatar from "@material-ui/core/Avatar";
import Chip from "@material-ui/core/Chip";
import ArrowUp from "@material-ui/icons/KeyboardArrowUp";
import ArrowDown from "@material-ui/icons/KeyboardArrowDown";
import IconButton from "@material-ui/core/IconButton";
import { DISH_PHOTOS_URL } from "shared-components/constants";

const styles = theme => ({
  list: {
    width: "100%",
    maxWidth: 360,
    maxHeight: 200,
    overflow: "auto",
    backgroundColor: theme.palette.background.paper
  },
  listItems: {
    display: "flex",
    flexDirection: "column"
  },
  margin: {
    margin: theme.spacing.unit
  },
  arrows: {
    display: "flex",
    flexDirection: "column",
    padding: "0px"
  },
  iconButton: {
    padding: "0px"
  },
  badge: {
    top: 1,
    left: -15
  }
});

class CheckboxListSecondary extends React.Component {
  state = {
    checked: [1]
  };

  render() {
    const {
      classes,
      items,
      onDishUp,
      onDishDown,
      onDishDelete,
      onDishClick
    } = this.props;

    return (
      <div className={classes.list}>
        <List dense className={classes.listItems}>
          {items.map(({ name, id, count, size_id, defaultPhoto }) => (
            <ListItem key={id}>
              <div className={classes.arrows}>
                <IconButton
                  className={classes.iconButton}
                  aria-label="Add"
                  onClick={() => {
                    onDishUp(id);
                  }}
                >
                  <ArrowUp />
                </IconButton>
                <IconButton
                  className={classes.iconButton}
                  aria-label="Remove"
                  onClick={() => {
                    onDishDown(id);
                  }}
                >
                  <ArrowDown />
                </IconButton>
              </div>
              <Badge badgeContent={count} color="primary">
                <Chip
                  avatar={
                    <Avatar
                      alt="Natacha"
                      src={`${DISH_PHOTOS_URL + defaultPhoto}`}
                    />
                  }
                  label={name}
                  onDelete={() => onDishDelete(id)}
                  className={classes.chip}
                  clickable
                  variant="outlined"
                  onClick={() => onDishClick(id)}
                />
              </Badge>
            </ListItem>
          ))}
        </List>
      </div>
    );
  }
}

CheckboxListSecondary.propTypes = {
  classes: PropTypes.object.isRequired
};

export default withStyles(styles)(CheckboxListSecondary);
