// components/pages/checkout-page/add-card/promo-code/promo-code.js
import React, { Fragment, Component } from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import styles from "../styles";
import Typography from "@material-ui/core/Typography";
import { Button, TextField } from "@material-ui/core";
import * as R from "ramda";
import { getPromotionByCode } from "api";

class RestaurantPage extends Component {
  state = {
    promoCode: "",
    notValid: false,
    valid: false
  };
  handleClick = () => {
    getPromotionByCode(this.state.promoCode).then(res => {
      if (R.isNil(res)) {
        this.setState({ valid: false, notValid: true });
        setTimeout(() => {
          this.setState({ notValid: false });
        }, 2000);
      } else {
        this.setState({ valid: true, notValid: false });
        this.props.onValidPromocode(res);
      }
    });
  };

  handleChange = name => event => {
    this.setState({ [name]: event.target.value });
  };
  render() {
    const { classes } = this.props;
    const { promoCode, notValid, valid } = this.state;
    const validMessage = "You have a discount!";
    const notValidMessage = "This is not valid code";
    let message;
    if (valid)
      message = (
        <Typography gutterBottom variant="subtitle1" color="primary">
          {validMessage}
        </Typography>
      );
    else if (notValid)
      message = (
        <Typography gutterBottom variant="subtitle1" color="error">
          {notValidMessage}
        </Typography>
      );

    return (
      <Fragment>
        <div className={classes.promoCode}>
          <Typography gutterBottom variant="h6" component="h2">
            Promocode
          </Typography>
          <div className={classes.promoCodeContent}>
            <TextField
              style={{ minHeight: "3.35rem" }}
              className="input"
              variant="outlined"
              type="text"
              value={promoCode}
              onChange={this.handleChange("promoCode")}
              placeholder="Enter Promo Code"
            />
            <Button
              color="secondary"
              variant="outlined"
              disabled={R.isEmpty(promoCode)}
              size="large"
              style={{ minHeight: "3.35rem" }}
              onClick={this.handleClick}
            >
              Submit
            </Button>
          </div>
          {message}
        </div>
      </Fragment>
    );
  }
}

RestaurantPage.propTypes = {
  classes: PropTypes.object.isRequired
};

export default withStyles(styles)(RestaurantPage);
