// components/pages/dishes-page/index.js
import React, { Fragment, Component } from "react";

import Header from "components/shared-components/header";
import GridList from "components/grid-list";
import SearchPanel from "components/search-panel";
import RestaurantFilter from "components/restaurant-filter";
import DishPage from "../dish-page";
import DishListItem from "components/grid-list-item-dish";
import Spinner from "components/spinner";

import "./restaurant-page.css";

import { getAllDishes } from "api";

export default class DishesPage extends Component {
  state = {
    openDish: false,
    dishToShow: {},
    term: ""
  };

  componentDidMount() {
    getAllDishes().then(dishes => {
      this.setState({
        dishes
      });
    });
  }
  onSearchChange = term => {
    this.setState({ term });
  };
  onDishAddClick = dish => {
    this.setState({ dishToShow: dish, openDish: true });
  };

  search(dishes, term) {
    if (term.length === 0) {
      return dishes;
    }
    return dishes.filter(d => {
      return d.DishName.toLowerCase().indexOf(term.toLowerCase()) > -1;
    });
  }

  render() {
    const { dishes, term } = this.state;

    const visibleItems = this.search(dishes, term);

    const content = !dishes ? (
      <Spinner /> // loading
    ) : visibleItems && visibleItems.length ? ( // show result
      <GridList
        data={visibleItems}
        GridListItem={DishListItem}
        onDishAddClick={this.onDishAddClick}
      />
    ) : (
      // no result
      <span className="message">No results that match your search</span>
    );
    return (
      <Fragment>
        <Header
            user={this.props.user}
            handleClose={this.props.handleClose}
        />
        <div className="rest-page-content">
          <DishPage
            openDish={this.state.openDish}
            dishToShow={this.state.dishToShow}
          />
          <div className="filter-bar">
            <div className="search-bar">
              <SearchPanel
                placeholder="Search dishes"
                onSearchChange={this.onSearchChange}
              />
            </div>
            <RestaurantFilter />
          </div>
        </div>
        {content}
      </Fragment>
    );
  }
}
