// components/schedule-item-dishes-list/item-list2_DEL.js
import React from 'react';
import PropTypes from 'prop-types';
import Badge from '@material-ui/core/Badge';
import { withStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import ListItemText from '@material-ui/core/ListItemText';
import Checkbox from '@material-ui/core/Checkbox';
import Avatar from '@material-ui/core/Avatar';
import Chip from '@material-ui/core/Chip';
import RemoveIcon from '@material-ui/icons/Remove';
import AddIcon from '@material-ui/icons/Add';
import ArrowUp from '@material-ui/icons/KeyboardArrowUp';
import Divider from '@material-ui/core/Divider';

const styles = theme => ({
  list: {
    width: '100%',
    maxWidth: 360,
    maxHeight: 200,
    backgroundColor: theme.palette.background.paper,
  },
  margin: {
   margin: theme.spacing.unit,
 },
});

class CheckboxListSecondary extends React.Component {
  state = {
    checked: [1],
  };

  handleToggle = value => () => {
    const { checked } = this.state;
    const currentIndex = checked.indexOf(value);
    const newChecked = [...checked];

    if (currentIndex === -1) {
      newChecked.push(value);
    } else {
      newChecked.splice(currentIndex, 1);
    }

    this.setState({
      checked: newChecked,
    });
  };

  render() {///
    const { classes, items } = this.props;
    const itemsToShow = items ? items : [];
      return (
        <div className={classes.list}>
          <List dense>
            {itemsToShow.map(({name, id:value, count}) => (
              <ListItem key={value}>
                <ArrowUp />
                <Chip
                       avatar={<Badge  badgeContent={count} color="primary">
                                <Avatar alt="Natacha" src={`/images/dishes/${value}.jpg`} />

                               </Badge>}
                       label={name}
                       onDelete={this.handleToggle}
                       className={classes.chip}
                       clickable
                     />
              </ListItem>
            ))}
          </List>
        </div>
      );
  }
}

CheckboxListSecondary.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(CheckboxListSecondary);
