// components/pages/schedule-items-page/schedule-items-page.js
import React, { Component } from "react";
import BigCalendar from "react-big-calendar";
import "react-big-calendar/lib/css/react-big-calendar.css";
import moment from "moment";
import ScheduleDialog from "./schedule-dialog";

const localizer = BigCalendar.momentLocalizer(moment);
const events = [];

export default class MyCalendar extends Component {
  state = {
    events,
    selectedEvent: {},
    dialogOpen: false,
    editItem: false
  };

  handleAddApply = event => {
    this.setState({
      events: [...this.state.events, event]
    });
  };
  handleEditApply = newEvent => {
    const oldEvents = this.state.events;
    const idx = oldEvents.findIndex(
      ev => ev.start === newEvent.start && ev.end === newEvent.end
    );

    this.setState({
      events: [
        ...oldEvents.slice(0, idx),
        newEvent,
        ...oldEvents.slice(idx + 1)
      ]
    });
  };

  handleDelete = eventToDel => {
    const oldEvents = this.state.events;
    const idx = oldEvents.findIndex(
      ev => ev.start === eventToDel.start && ev.end === eventToDel.end
    );

    this.setState({
      events: [...oldEvents.slice(0, idx), ...oldEvents.slice(idx + 1)]
    });
  };

  handleSelectSlot = event => {
    this.setState({
      dialogOpen: true,
      editItem: false,
      selectedEvent: { ...event, dishes: [] }
    });
  };
  handleSelectEvent = event => {
    this.setState({
      dialogOpen: true,
      editItem: true,
      selectedEvent: event
    });
  };
  render(props) {
    const handleApply = this.state.editItem
      ? this.handleEditApply
      : this.handleAddApply;
    return (
      <div>
        <ScheduleDialog
          dialogOpen={this.state.dialogOpen}
          event={this.state.selectedEvent}
          handleApply={handleApply}
          handleDelete={this.handleDelete}
          editItem={this.state.editItem}
        />
        <BigCalendar
          style={{ height: "420px", marginLeft: "5rem", marginRight: "5rem" }}
          selectable
          popup
          showMultiDayTimes
          localizer={localizer}
          events={this.state.events}
          defaultView={BigCalendar.Views.WEEK}
          scrollToTime={new Date(1970, 1, 1, 6)}
          onSelectSlot={this.handleSelectSlot}
          onSelectEvent={this.handleSelectEvent}
        />
      </div>
    );
  }
}
