// components/schedule-item-dish-sizes-list/schedule-item-dish-sizes-list.js
import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormControl from '@material-ui/core/FormControl';

const styles = theme => ({
  root: {
    display: 'flex',
  },
  formControl: {
    margin: theme.spacing.unit * 1,
  },
  group: {
    margin: `${theme.spacing.unit}px 0`,
  },
});



class DishSizesList extends React.Component {
  state = {
   selected: 0,
 };

 handleChange = event => {
   console.log("TARGET", event.target);
   this.setState({ selected: event.target.value });
   this.props.onDishSizeChange(event.target.value);
 };
 componentDidMount(nextProps) {
   console.log("sizes did mount");
    this.setState({ selected: this.props.selected });
 }
  render() {
    const { classes, items } = this.props;
    const itemsToShow = items ? items : [];
      return (
        <div className={classes.list}>
          <FormControl component="fieldset" className={classes.formControl}>
               <RadioGroup
                 aria-label="Select size"
                 name="Dish sizes"
                 className={classes.group}
                 value={this.state.selected.toString()}
                 onChange={this.handleChange}
               >
                 {itemsToShow.map(({id, name}) => (
                   <FormControlLabel key={id} value={id.toString()} control={<Radio />} label={name} />
                 ))}
                   />
               </RadioGroup>
           </FormControl>
        </div>
      );
  }
}

DishSizesList.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(DishSizesList);
