// components-illia/pages/login-page/login-form/login-form.js
import React, { Component } from "react";
import { Link } from "react-router-dom";

import Paper from "@material-ui/core/Paper";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogContentText from "@material-ui/core/DialogContentText";
import DialogTitle from "@material-ui/core/DialogTitle";
import "../login-page.css";
import { Redirect } from "react-router-dom";
import client from "feathers/client";
import Spinner from "shared-components/spinner";


export default class LoginForm extends Component {
  state = {
    open: false,
    isLoggedIn: false,
    email: '',
    password: ''
  };
  async login() {
    const { email, password } = this.state;
    this.setState({ loading: true });
    await client
      .authenticate({
        strategy: "local",
        email,
        password
      })
      .then(response => client.passport.verifyJWT(response.accessToken))
      .then(payload => client.service("customers").get(payload.userId))
      .then(user => {
        client.set("user", user);
      })
      .then(() => {
        this.setState({ error: null, isLoggedIn: true, loading: false });
      })
      .catch(error => {
        this.setState({ error, errlog: true, errpass: true, loading: false });
      });
  }

  updateField(name, ev) {
    this.setState({ [name]: ev.target.value });
  }

  handleClickOpen = () => {
    this.setState({ open: true });
  };

  handleClose = () => {
    this.setState({ open: false });
  };
  render() {
    console.log("IS LOADING", this.state.loading);
    if (this.state.loading) return <Spinner />;
    if (!this.state.loading && this.state.isLoggedIn)
      return <Redirect to="/" />;
    return (
      <div className="form">
        <Paper className="paper-form" elevation={15}>
          <span className="label">Log in</span>
          <p>{this.state.error && this.state.error.message}</p>
          <TextField
            className="input"
            label="Email"
            margin="normal"
            variant="outlined"
            placeholder="illia.sydun@gmail.com"
            error={this.state.errlog}
            value={this.state.email}
            onFocus={()=>{this.setState({errlog:false})}}
            onChange={ev => this.updateField("email", ev)}
          />
          <TextField
            className="input"
            label="Password"
            margin="normal"
            variant="outlined"
            type="password"
            placeholder="passwordpass"
            error={this.state.errpass}
            value={this.state.password}
            onFocus={()=>{this.setState({errpass:false})}}
            onChange={ev => this.updateField("password", ev)}
          />
          <div className="buttons">
            <Button className="button" onClick={() => this.login()}>
              Log in
            </Button>
            <a href="localhost:3000" onClick={this.handleClickOpen}>
              Forgot your password?
            </a>
            <Link to="/registration">Don't have one?</Link>
          </div>

          <Dialog
            open={this.state.open}
            onClose={this.handleClose}
            aria-labelledby="form-dialog-title"
          >
            <DialogTitle id="form-dialog-title">Reset password</DialogTitle>
            <DialogContent>
              <DialogContentText>
                To reset your password, please enter your email address here. We
                will send you confirmation link.
              </DialogContentText>
              <TextField
                autoFocus
                margin="dense"
                id="name"
                label="Email Address"
                type="email"
                fullWidth
              />
            </DialogContent>
            <DialogActions>
              <Button className="button" onClick={this.handleClose}>
                Cancel
              </Button>
              <Button className="button" onClick={this.handleClose}>
                Confirm
              </Button>
            </DialogActions>
          </Dialog>
        </Paper>
      </div>
    );
  }
}
