// components/pages/dish-page/dish-page.js
import React from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import SwipeableDrawer from "@material-ui/core/SwipeableDrawer";
import DishSidePage from "./dish-side-page";

const styles = {
  list: {
    width: 250
  }
};

class SwipeableTemporaryDrawer extends React.Component {
  state = {
    right: false
  };
  toggleDrawer = (side, open) => () => {
    this.setState({
      [side]: open
    });
    if (this.props.handleClose) this.props.handleClose();
  };
  handleApply = () => {
    this.setState({
      right: false
    });
    if (this.props.handleClose) this.props.handleClose();
  };
  componentWillReceiveProps(nextProps) {
    if (nextProps.openDish !== this.state.right) {
      this.setState({
        right: nextProps.openDish
      });
    }
  }
  render() {
    return (
      <div>
        <SwipeableDrawer
          anchor="right"
          open={this.state.right}
          onClose={this.toggleDrawer("right", false)}
          onOpen={this.toggleDrawer("right", true)}
        >
          <div tabIndex={0} role="button">
            <DishSidePage
              dishToShow={this.props.dishToShow}
              handleApply={this.handleApply}
              updateDish={this.props.updateDish}
            />
          </div>
        </SwipeableDrawer>
      </div>
    );
  }
}

SwipeableTemporaryDrawer.propTypes = {
  classes: PropTypes.object.isRequired
};

export default withStyles(styles)(SwipeableTemporaryDrawer);
