// components-illia/pages/account-page/account-settings/account-settings.js
import React, { Component } from "react";

import Paper from "@material-ui/core/Paper";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import "../account-page.css";
import { updateUser } from "api";
import {authStore} from "store";
import {Typography} from "@material-ui/core";

export default class AccountSettings extends Component {
  constructor() {
    super();
    const customer = authStore.user;
    console.log(customer);
    this.state = {
      customer: {
        oldName: customer.CustName,
        oldSurname: customer.CustSurname,
        oldEmail: customer.email,
        oldPhone: customer.CustPhone
      }
    };
  }

  updateCustomer = () => {
    const { newCustomer } = this.state;
    updateUser(newCustomer);
  };
  render() {
    const { oldName, oldSurname, oldEmail, oldPhone } = this.state.customer;

    console.log(this.state);
    return (
      <Paper className="account-setting" elevation={15}>
          <Typography gutterBottom variant="h6" component="h2">
              Account settings
          </Typography>
        <div className="setting-form">
          <TextField
            className="input"
            label="Name"
            margin="normal"
            variant="outlined"
            type="text"
            value={
              this.state.newCustomer && this.state.newCustomer.CustName
                ? this.state.newCustomer.CustName
                : oldName
            }
            onChange={e =>
              this.setState({
                newCustomer: {
                  ...this.state.newCustomer,
                  CustName: e.target.value
                }
              })
            }
          />
          <TextField
            className="input"
            label="Surname"
            margin="normal"
            variant="outlined"
            type="text"
            value={
              this.state.newCustomer && this.state.newCustomer.CustSurname
                ? this.state.newCustomer.CustSurname
                : oldSurname
            }
            onChange={e =>
              this.setState({
                newCustomer: {
                  ...this.state.newCustomer,
                  CustSurname: e.target.value
                }
              })
            }
          />
          <TextField
            className="input"
            label="Email"
            margin="normal"
            variant="outlined"
            type="email"
            value={
              this.state.newCustomer && this.state.newCustomer.email
                ? this.state.newCustomer.email
                : oldEmail
            }
            onChange={e =>
              this.setState({
                newCustomer: {
                  ...this.state.newCustomer,
                  email: e.target.value
                }
              })
            }
          />
          <TextField
            className="input"
            label="Phone Nubmer"
            margin="normal"
            variant="outlined"
            type="text"
            value={
              this.state.newCustomer && this.state.newCustomer.CustPhone
                ? this.state.newCustomer.CustPhone
                : oldPhone
            }
            onChange={e =>
              this.setState({
                newCustomer: {
                  ...this.state.newCustomer,
                  CustPhone: e.target.value
                }
              })
            }
          />
          <TextField
            className="input"
            label="Password"
            margin="normal"
            variant="outlined"
            type="password"
            defaultValue="illia.sydun"
          />
        </div>
        <div className="setting-form-buttons">
          <Button className="button" onClick={this.updateCustomer}>
            Save
          </Button>
        </div>
      </Paper>
    );
  }
}
