// components/dish-side-page/styles.js
import {
  createStyles,
} from "@material-ui/core/styles";
  export default theme =>
  createStyles({
    list: {
      width: "383px",
      overflow: "hidden",
      marginRight: "-17px"
    },
    topDesc: {
      textAlign: "center",
      marginRight: "8px"
    },
    text: {
      paddingLeft: theme.spacing.unit * 2
    },
    textField: {
      marginLeft: theme.spacing.unit,
      textAlign: "center",
      width: "95%"
    },
    textNumField: {
      marginLeft: theme.spacing.unit,
      width: "2rem",
      webkitAppearance: "none"
    },
    flexRow: {
      display: "flex",
      marginLeft: theme.spacing.unit,
      flexDirection: "row",
      alignItems: "center",
      justifyContent: "space-between"
    },
    flexPrice: {
      width: "4.5rem",
      marginLeft: 0
    },
    price: {
      display: "flex",
      alignItems: "center",
      marginRight: theme.spacing.unit * 2,
      padding: theme.spacing.unit * 1
    },
    button: {
      width: "96%",
      marginLeft: theme.spacing.unit
    },
    bottomPanel: {
      width: "100%",
      position: "absolute",
      bottom: 0,
      marginBottom: theme.spacing.unit * 1
    },
    content: {
      overflowY: "scroll",
      overflowX: "hidden",
      paddingRight: "17px",
      width: "100%",
      bottom: "5rem",
      maxHeight: "86vh"
    },
    dishImg: {
      height: '34vh'
    }
  });
