// components/grid-list-item-dish/index.js
import React from 'react';

import GridListTileBar from '@material-ui/core/GridListTileBar';
import Button from '@material-ui/core/Button';
import PropTypes from 'prop-types';
import { withStyles,createStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import * as R from 'ramda'

import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles';

import './grid-list-item.css'

const theme = createMuiTheme({
  palette: {
    primary: {
       main: '#FAFAFA',
     },
     secondary: {
       main: '#BDBDBD',
     }, // This is just green.A700 as hex.
  },
});


const styles = theme =>
createStyles({
  button: {
    margin: theme.spacing.unit,
    borderColor: 'white'
  },
  tileBar: {
    height: "auto",
    padding: "0.5rem"
  },
  priceBar: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center"
  },
  extendedIcon: {
    marginRight: theme.spacing.unit,
  },
});



const GridListItem = ({id, name, price, description, restaurant, classes}) => {

  const shortDescription = `${R.take(140, description)}${description.length > 140 ? '...' : ''}` ;

  return (
  <div className='list-item'>
    <img src={`/images/dishes/${id}.jpg`} alt={name} />
    <MuiThemeProvider theme={theme}>
          <GridListTileBar
        className={classes.tileBar}
        title={
                <div>
                    <Typography variant="subtitle2" color="secondary" gutterBottom noWrap>
                      "{restaurant}" restaurant
                    </Typography>
                    <Typography variant="title" color="primary" gutterBottom noWrap>
                      {name}
                    </Typography>
                </div>
              }
        subtitle={
                  <div className="description">
                    {shortDescription}
                  </div>
                  }
        actionIcon={
                    <div className={classes.priceBar}>
                      <Typography variant="subtitle1" color="primary">
                          {`$${price}`}
                       </Typography>
                      <Button  variant="outlined" aria-label="Add" color="primary" className={classes.button}>
                      Add
                      </Button>
                    </div>
                    }
      />
    </MuiThemeProvider>
</div>
  )
}

GridListItem.propTypes = {
  classes: PropTypes.object.isRequired,
};


export default withStyles(styles)(GridListItem);
