// components/restaurant-filter/index.js
import React from 'react';

import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';

import SeatsPicker from './seats-picker';
import DatePicker from './date-picker';
import TimePicker from './time-picker';
import CountryPicker from './country-picker';
import CityPicker from './city-picker';
import CuisinePicker from './cuisine-picker';
import './table-book-bar.css'

import FilterDialog from 'components/filter-dialog';

const dialogContent = (
      <DialogContent className='dialog-elements'>
        <DialogTitle id="form-dialog-title">Seats number</DialogTitle>
        <SeatsPicker />
        <DialogTitle id="form-dialog-title">Date & Time</DialogTitle>
        <DatePicker />
        <TimePicker />
        <DialogTitle id="form-dialog-title">Location</DialogTitle>
        <CountryPicker />
        <CityPicker />
        <DialogTitle id="form-dialog-title">Cuisine</DialogTitle>
        <CuisinePicker />
      </DialogContent>
    );


export default class RestaurantFilter extends React.Component {

  state = {
        noFilter: false,
        country: '',
        city: '',
        seatsNumber: '',
        cuisine: ''
  }

  onFilterApply = () => {
    console.log("filter applied" , this.state);
    this.props.onFilterChange(this.state);
  }

  onCountryChange = (value) => {
    console.log("country changed", value);
  this.setState({
    country: value
  })
};
  onCityChange = (value) => {
    this.setState({
      city: value
    })
  }
  onSeatsChange = (value) => {
    this.setState({
      seats: value
    })
  }
  onCuisineChange = (value) => {
    this.setState({
      cuisine: value
    })
  }

  render() {

    return (
      <FilterDialog
        btnName='Filters'
        onFilterApply = { this.onFilterApply }
        dialogContent = {
          <DialogContent className='dialog-elements'>
            <DialogTitle id="form-dialog-title">Seats number</DialogTitle>
            <SeatsPicker onSelectChange={ this.onSeatsChange } />
            <DialogTitle id="form-dialog-title">Date & Time</DialogTitle>
            <DatePicker />
            <TimePicker />
            <DialogTitle id="form-dialog-title">Location</DialogTitle>
            <CountryPicker onSelectChange={  this.onCountryChange }/>
            <CityPicker  onSelectChange={ this.onCityChange }/>
            <DialogTitle id="form-dialog-title">Cuisine</DialogTitle>
            <CuisinePicker onSelectChange={ this.onCuisineChange }/>
          </DialogContent>
        } />
    );
  }
}
