// shared-components/radio-select/radio-select.js
import React from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import Radio from "@material-ui/core/Radio";
import RadioGroup from "@material-ui/core/RadioGroup";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import FormControl from "@material-ui/core/FormControl";

const styles = theme => ({
  root: {
    display: "flex"
  },
  formControl: {
    marginLeft: theme.spacing.unit * 1,
    marginTop: 0
  },
  group: {
    marginLeft: theme.spacing.unit * 1,
    marginTop: 0,
    flexDirection: "column"
  }
});

class RadioSelect extends React.Component {
  state = {
    selected: 1
  };

  handleChange = event => {
    this.setState({ selected: event.target.value });
    this.props.onChange(event.target.value);
  };
  componentDidMount(nextProps) {
    if (this.props.selected) this.setState({ selected: this.props.selected });
  }
  render() {
    const { classes, items, direction } = this.props;
    const itemsToShow = items ? items : [];
    const groupDirection = direction ? direction : classes.group.flexDirection;

    return (
      <div className={classes.list}>
        <FormControl component="fieldset" className={classes.formControl}>
          <RadioGroup
            aria-label="Select size"
            name="Dish sizes"
            className={classes.group}
            style={{ flexDirection: groupDirection }}
            value={this.state.selected.toString()}
            onChange={this.handleChange}
          >
            {itemsToShow.map(({ id, name }) => (
              <FormControlLabel
                key={id}
                value={id.toString()}
                control={<Radio />}
                label={name}
              />
            ))}
            />
          </RadioGroup>
        </FormControl>
      </div>
    );
  }
}

RadioSelect.propTypes = {
  classes: PropTypes.object.isRequired
};

export default withStyles(styles)(RadioSelect);
