// components/pages/restaurants-page/restaurant-filter/index.js
import React from "react";

import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";
import DatePicker from "./date-picker";
import TimePicker from "./time-picker";
import "./table-book-bar.css";
import FilterDialog from "../filter-dialog";
import SelectControl from "shared-components/select-control";
import {
  citiesSelectItems,
  countriesSelectItems,
  cuisinesSelectItems,
  seatsSelectItems
} from "api";

export default class RestaurantFilter extends React.Component {
  state = {
    noFilter: false,
    country: "",
    city: "",
    seats: "",
    cuisine: ""
  };

  onFilterApply = () => {
    this.props.onFilterChange(this.state);
  };

  onSelectChange = ({ name, value }) => {
    this.setState({
      [name]: value
    });
  };

  onDialogClose = () => {
    this.setState(this.props.filter);
  };

  render() {
    return (
      <FilterDialog
        btnName="Filters"
        onFilterApply={this.onFilterApply}
        onDialogClose={this.onDialogClose}
        dialogContent={
          <DialogContent className="dialog-elements">
            <DialogTitle id="form-dialog-title">Seats number</DialogTitle>
            <SelectControl
              onSelectChange={this.onSelectChange}
              name="seats"
              items={seatsSelectItems}
              value={this.state.seats}
              emptyValue="Select seats number"
            />
            <DialogTitle id="form-dialog-title">Date & Time</DialogTitle>
            <DatePicker />
            <TimePicker />
            <DialogTitle id="form-dialog-title">Location</DialogTitle>
            <SelectControl
              onSelectChange={this.onSelectChange}
              name="country"
              items={countriesSelectItems}
              value={this.state.country}
              emptyValue="Select country"
            />
            <SelectControl
              onSelectChange={this.onSelectChange}
              name="city"
              items={citiesSelectItems}
              value={this.state.city}
              emptyValue="Select city"
            />
            <DialogTitle id="form-dialog-title">Cuisine</DialogTitle>
            <SelectControl
              onSelectChange={this.onSelectChange}
              name="cuisine"
              items={cuisinesSelectItems}
              value={this.state.cuisine}
              emptyValue="Select cuisine"
            />
          </DialogContent>
        }
      />
    );
  }
}
