// components/pages/restaurants-page/index.js
import React, { Fragment, Component } from 'react';

import Header from 'components/shared-components/header';
import GridList from 'components/grid-list';
import SearchPanel from 'components/search-panel';
import RestaurantFilter from 'components/restaurant-filter';

import RestListItem from 'components/grid-list-item-restaurant';

import './restaurant-page.css'

import restaurants from 'api/mock-restaurants';

const navItems = [
  {label: 'Dishes', link: 'dishes' },
  {label: 'Login', link: 'login' },
  {label: 'Registration', link: 'registration'}
];


export default class RestaurantPage extends Component{

  state = {
    restaurants: restaurants,
    term: '',
    filter: {
              noFilter: false,
              country: '',
              city: '',
              seats: '',
              cuisine: ''
              }
  }

  onSearchChange = (term) => {
  this.setState({ term });
};
  onFilterChange = (filter) => {
  this.setState({ filter });
};
  search(restaurants, term) {
    if(term.length === 0) {
      return restaurants;
    }
    return restaurants.filter( (r) => {
      return r.name
             .toLowerCase()
             .indexOf(term.toLowerCase()) > -1
    })
  };

  filter(restaurants, filter ) {
    if(filter.noFilter) return restaurants;

    let filtered = restaurants;

    if(filter.country.length !== 0){
      filtered = filtered.filter((r) => r.country === filter.country);
    }
    if(filter.city.length !== 0){
      filtered = filtered.filter((r) => r.city === filter.city);
    }
    if(filter.seats.length !== 0){
      filtered = filtered.filter((r) => r.seats >= filter.seats);
    }
    if(filter.cuisine.length !== 0){
      filtered = filtered.filter((r) => r.cuisine === filter.cuisine);
    }
    return filtered;
  }

  render(){
    const { restaurants, filter, term } = this.state;

    const visibleItems = () => this.filter(
                        this.search(restaurants, term),
                        filter);
    const DataToShow = visibleItems().length !== 0 ? () => <GridList data={ visibleItems() } GridListItem = {RestListItem}/>
                                                   : () => <span className="message">No results that match your search</span>;

    return (
      <Fragment>
        <Header navItems={ navItems } />
        <div className='rest-page-content'>
          <div className="filter-bar">
            <div className="search-bar">
              <SearchPanel onSearchChange = {this.onSearchChange} placeholder="Search restaurants"/>
            </div>
              <RestaurantFilter onFilterChange={this.onFilterChange} filter={this.state.filter}/>
          </div>
        </div>
              <DataToShow />
      </Fragment>
    )
  };
}
