// components/pages/restaurants-page/index.js
import React, { Fragment, Component } from 'react';

import Header from 'components/shared-components/header';
import GridList from 'components/grid-list';
import SearchPanel from 'components/search-panel';
import RestaurantFilter from 'components/restaurant-filter';
import GridListTile from '@material-ui/core/GridListTile';
import RestListItem from 'components/grid-list-item-restaurant'

import './restaurant-page.css'

import restaurants from 'api/mock-restaurants';

const navItems = [
  {label: 'Join us'},
  {label: 'Contact'},
  {label: 'Login'},
  {label: 'Registration'}
];


export default class RestaurantPage extends Component{

  state = {
    restaurants: restaurants,
    term: '',
    filter: {
              noFilter: true,
              country: '',
              city: '',
              seatsNumber: '',
              cuisine: ''
              }
  }

  onSearchChange = (term) => {
  this.setState({ term });
};
  onFilterChange = (filter) => {
  this.setState({ filter });
};

  search(restaurants, term) {
    if(term.length === 0) {
      return restaurants;
    }

    return restaurants.filter( (r) => {
      return r.name
             .toLowerCase()
             .indexOf(term.toLowerCase()) > -1
    })
  };

  filter(restaurants, filter) {
    if(filter.noFilter) return restaurants;

    if(filter.country.length !== 0){
      restaurants.filter((r) => r.country = filter.country);
    }
    if(filter.city.length !== 0){
      restaurants.filter((r) => r.city = filter.city);
    }
    if(filter.seatsNumber.length !== 0){
      restaurants.filter((r) => r.seatsNumber = filter.seatsNumber);
    }
    if(filter.cuisine.length !== 0){
      restaurants.filter((r) => r.cuisine = filter.cuisine);
    }
    return restaurants;
  }

  render(){
    const { restaurants, filter, term } = this.state;

    const visibleItems = this.filter(
                        this.search(restaurants, term),
                        filter);

    return (
      <Fragment>
        <Header navItems={ navItems } />
        <div className='rest-page-content'>
          <div className="filter-bar">
            <div className="search-bar">
              <SearchPanel onSearchChange = {this.onSearchChange}/>
            </div>
              <RestaurantFilter onFilterChange={this.onFilterChange}/>
          </div>
        </div>
        <GridList data={ visibleItems } GridListItem = {RestListItem}/>
      </Fragment>
    )
  };
}
