// api/addresses.js
import client from "feathers/client";
import { authStore } from "store";

export const getCustomerDeliveryAddresses = () =>
  client
    .service("customer-delivery-addresses")
    .find({ query: { Cust_id: authStore.user.id } })
    .then(res =>
      Promise.all(
        res.data.map(async addr => {
          addr.address.id = addr.id;
          addr.address.Address_id = addr.Address_id;
          addr.address.CityName = await getCity(addr.address.City_id).then(
            City => City.CityName
          );
          return addr.address;
        })
      )
    )
    .then(addresses => addresses);
export const createDeliveryAddresses = address =>
  createAddress(address).then(({ id }) =>
    client
      .service("customer-delivery-addresses")
      .create({ Cust_id: authStore.user.id, Address_id: id })
  );

export const createAddress = address =>
  client.service("addresses").create(address);

export const getCity = City_id => client.service("cities").get(City_id);
export const getAllCities = () =>
  client
    .service("cities")
    .find()
    .then(res => res.data);
export const getCitiesByCountryId = Country_id =>
  client
    .service("cities")
    .find({ query: { Country_id } })
    .then(res => res.data);

export const getAllCountries = () =>
  client
    .service("countries")
    .find()
    .then(res => res.data);
