// api/addresses.js
import client from "feathers/client";

export const getCustomerDeliveryAddresses = () =>
    client
        .service("customer-delivery-addresses")
        .find({ query: { Cust_id: client.get("user").id } })
        .then(res =>
            Promise.all(
                res.data.map(async addr => {
                    addr.address.id = addr.id;
                    addr.address.CityName = await getCity(addr.address.City_id).then(
                        City => City.CityName
                    );
                    return addr.address;
                })
            )
        )
        .then(addresses => addresses);
export const createDeliveryAddresses = address =>
    client
        .service("addresses")
        .create(address)
        .then(Address_id => {
            client
                .service("customer-delivery-addresses")
                .create({ Cust_id: client.get("user").id, Address_id });
        });

export const getCity = City_id => client.service("cities").get(City_id);
export const getAllCities = () => client.service("cities").find();
export const getCitiesByCountryId = Country_id =>
    client.service("cities").find({ query: { Country_id } });

export const getAllCountries = () => client.service("countries").find();
