// api/dishes.js
import client from "feathers/client";

export const getAllDishes = () =>
    client
        .service("restaurant-dishes")
        .find()
        .then(res => res.data);

export const getDishesByRestId = Restaurant_id =>
    client
        .service("restaurant-dishes")
        .find({ query: { Restaurant_id } })
        .then(res => res.data);

export const getDish = Dish_id =>
    client.service("restaurant-dishes").findOne({ query: { Dish_id } });


export const getDishInMenu = MenuDish_id =>
    client.service("dishes-in-menu").findOne({ query: { MenuDish_id } });

export const getDishesByDishCategoryId = Category_id =>
    getAllDishes().then(res =>
        res.filter(resDish => resDish.Category_id === Category_id)
    );

export const getCategoriesByMenuId = menuId =>
    client
        .service("menu-categories")
        .find({ query: { Menu_id: menuId } })
        .then(res => res.data.menu_categories);

export const getDishCategoriesByRestId = Restaurant_id =>
    client
        .service("menu")
        .findOne({ query: { Restaurant_id } })
        .then(menu => menu.menu_categories)
        .catch(() => []);
