// api/mock-restaurants.js

 export default [
  {
    id: 1,
    name: 'Progress',
    opened: '8:00 am',
    closed: '11:30 pm',
    country: 'Poland',
    city: 'Lublin',
    seats: '5'
  },
  {
    id: 2,
    name: 'Stylowa',
    opened: '8:30 am',
    closed: '11:30 pm',
    country: 'Poland',
    city: 'Lublin',
    seats: '5'
  },
  {
    id: 3,
    name: 'Pod Aniolami',
    opened: '8:30 am',
    closed: '11:40 pm',
    country: 'Poland',
    city: 'Lublin',
    seats: '5'
  },
  {
    id: 4,
    name: 'Pod Wawelem-Kompania Kuflow',
    opened: '8:30 am',
    closed: '01:30 am',
    country: 'Poland',
    city: 'Lublin',
    seats: '5'
  },
  {
   id: 5,
   name: 'Dawno Temu na Kazimierzu',
   opened: '8:30 am',
   closed: '11:30 pm',
   country: 'Poland',
   city: 'Warsaw',
   seats: '5'
  },
  {
   id: 6,
   name: 'Cyrano de Bergerac',
   opened: '8:30 am',
   closed: '11:30 pm',
   country: 'Poland',
   city: 'Warsaw',
   seats: '5'
  },
  {
   id: 7,
   name: 'Café Oranzeria',
   opened: '8:30 am',
   closed: '11:30 pm',
   country: 'Ukraine',
   city: 'Kiev',
   seats: '5'
  },
  {
    id: 8,
    name: 'Alebriche',
    opened: '8:30 am',
    closed: '11:30 pm',
    country: 'Ukraine',
    city: 'Kiev',
    seats: '5'
  },
  {
    id: 9,
    name: 'Portretowa',
    opened: '8:30 am',
    closed: '11:30 pm',
    country: 'Ukraine',
    city: 'Kiev',
    seats: '5'
  },
  {
    id: 10,
    name: 'Dawnie smaki',
    opened: '8:30 am',
    closed: '11:30 pm',
    country: 'Ukraine',
    city: 'Kiev',
    seats: '5'
  },
  {
    id: 11,
    name: 'Stary dom',
    opened: '8:30 am',
    closed: '11:30 pm',
    country: 'Ukraine',
    city: 'Kiev',
    seats: '5'
  }
      ];
