// api/mock-dishes.js

 export default [
  {
    id: 1,
    name: 'Wasabi Shrimp and Scallop',
    price: 11.95,
    category_id: [1, 3],
    description: 'Shrimp and scallops, green and sweet onion, cucumber, masago, and wasabi aioli',
    restaurant: 'Progress',
    size_id: [1, 2, 3]
  },
  {
    id: 2,
    name: 'Poutine Your Mouth',
    price: 10,
    category_id:[1, 3],
    description: 'Waffle fries topped with kimchee and furikake',
    restaurant: 'Stylowa',
    size_id: [1, 2]
  },
  {
    id: 3,
    name: 'Piadina Flatbread',
    price: 12,
    category_id:[1, 3],
    description: 'Your choice of Margherita, Il Rigoletto, Veneziano, or Piccola Italia',
    restaurant: 'Pod Aniolami',
    size_id: [2, 3]
  },
  {
    id: 4,
    name: 'Pollo Parmigiano',
    price: 16,
    category_id: [1, 4],
    description: 'Pan fried breaded chicken breast, fresh mozzarella, tomato sauce, and linguini pasta',
    restaurant: 'Pod Wawelem-Kompania Kuflow',
    size_id: [1, 2 , 3]
  },
  {
   id: 5,
   name: 'Tagliere di Salumi',
   price: 13,
   category_id: [1, 5],
   description: 'House charcuterie platter with prosciutto di parma, nresaola, coppa, soppressata, pecorino, and parmigiano reggiano',
   restaurant: 'Dawno Temu na Kazimierzu',
   size_id: [1, 2]
  },
  {
   id: 6,
   name: 'Ricotta Croccante',
   price: 12,
   category_id: [1, 5],
   description: 'Vegetarian. Grilled rustic bread, hand pressed ricotta, spicy red pepper, and truffle honey',
   restaurant: 'Cyrano de Bergerac',
   size_id: [2, 3]
  },
  {
   id: 7,
   name: 'Pollo Parno',
   price: 16,
   category_id: [1, 2],
   description: 'Pan fried breaded chicken breast, tomato sauce, and linguini pasta',
   restaurant: 'Café Oranzeria',
   size_id: [1, 2 ,3 ]
  },
  {
    id: 8,
    name: 'Capellini con Gamberetti',
    price: 18,
    category_id: [1, 2],
    description: 'Sauteed shrimp, rehydrated sun-dried tomatoes, Kalamata olives, caramelized onions, and baby spinach, Sauteed shrimp, rehydrated sun-dried tomatoes, Kalamata olives, caramelized onions, and baby spinach',
    restaurant: 'Alebriche',
    size_id: [1, 2, 3]
  },
  {
    id: 9,
    name: `Purè di Patate`,
    price: 8,
    category_id: [1, 6],
    description: 'Vegetarian. Garlic mashed potatoes.',
    restaurant: 'Portretowa',
    size_id: [1, 2]
  },
  {
    id: 10,
    name: 'Cajun Yellow Fin Tuna',
    price: 28,
    category_id: [1, 6],
    description: 'Pickled ginger salad, fennel, cucumber, lime vinaigrette, and cilantro',
    restaurant: 'Dawnie smaki',
    size_id: [1, 2, 3]
  }
      ];
