// api/mock-restaurants.js

 export default [
  {
    id: 1,
    name: 'Progress',
    opened: '8:00 am',
    closed: '11:30 pm',
    country: 'Poland',
    address: 'Warszawska 1a/10',
    city: 'Lublin',
    description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
    seats: '5',
    price: '$$$$'
  },
  {
    id: 2,
    name: 'Stylowa',
    opened: '8:30 am',
    closed: '11:30 pm',
    country: 'Poland',
    address: 'Narutowicza 20a',
    city: 'Lublin',
    description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
    seats: '5',
    price: '$$$$'
  },
  {
    id: 3,
    name: 'Pod Aniolami',
    opened: '8:30 am',
    closed: '11:40 pm',
    country: 'Poland',
    address: 'Muzyczna 101b',
    city: 'Lublin',
    description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
    seats: '5',
    price: '$$$$'
  },
  {
    id: 4,
    name: 'Pod Wawelem-Kompania Kuflow',
    opened: '8:30 am',
    closed: '01:30 am',
    country: 'Poland',
    address: 'Jana Pawla 56a',
    city: 'Lublin',
    description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
    seats: '5',
    price: '$$$'
  },
  {
   id: 5,
   name: 'Dawno Temu na Kazimierzu',
   opened: '8:30 am',
   closed: '11:30 pm',
   country: 'Poland',
   address: 'Widonska 5015a',
   city: 'Warsaw',
   description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
   seats: '5',
   price: '$$'
  },
  {
   id: 6,
   name: 'Cyrano de Bergerac',
   opened: '8:30 am',
   closed: '11:30 pm',
   country: 'Poland',
   address: 'Bolnska 1010b',
   city: 'Warsaw',
   description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
   seats: '5',
   price: '$$$$'
  },
  {
   id: 7,
   name: 'Café Oranzeria',
   opened: '8:30 am',
   closed: '11:30 pm',
   country: 'Ukraine',
   address: 'Creshchatyk 1a/506',
   city: 'Kiev',
   description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
   seats: '5',
   price: '$$$'
  },
  {
    id: 8,
    name: 'Alebriche',
    opened: '8:30 am',
    closed: '11:30 pm',
    country: 'Ukraine',
    address: 'Mostowa 45',
    city: 'Kiev',
    description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
    seats: '5',
    price: '$$$$'
  },
  {
    id: 9,
    name: 'Portretowa',
    opened: '8:30 am',
    closed: '11:30 pm',
    country: 'Ukraine',
    address: 'Koponalska 36',
    city: 'Kiev',
    description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
    seats: '5',
    price: '$$$'
  },
  {
    id: 10,
    name: 'Dawnie smaki',
    opened: '8:30 am',
    closed: '11:30 pm',
    country: 'Ukraine',
    address: 'Betropolska 33a',
    city: 'Kiev',
    description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
    seats: '5',
    price: '$$$$'
  },
  {
    id: 11,
    name: 'Stary dom',
    opened: '8:30 am',
    closed: '11:30 pm',
    country: 'Ukraine',
    address: 'Neprotoponska 106',
    city: 'Kiev',
    description: 'Very cozy and comfortable restaurant with tasty dishes and smilling staff',
    seats: '5',
    price: '$$$$'
  }
      ];
