// api/selects.js
import {
  getAllCities,
  getAllRestaurants,
  getCustomerDeliveryAddresses,
  getDishCategoriesByRestId,
  getDishesByDishCategoryId
} from "api";

export const dataToSelect = (arr, vProp, lProp) =>
  arr.map(item => ({
    value: item[vProp],
    label: item[lProp]
  }));

export const getCustomerAddressesSelect = () =>
  getCustomerDeliveryAddresses().then(address =>
    address.map(addr => ({
      value: addr.Address_id,
      label: `${addr.CityName}, ${addr.StreetName}, ${addr.StreetNumber}, ${
        addr.ZipCode
      }`
    }))
  );
export const getCitiesSelect = () =>
  getAllCities().then(cities =>
    cities.map(c => ({ value: c.id, label: c.CityName }))
  );

export const getRestaurantsSelect = () =>
  getAllRestaurants().then(restaurants =>
    restaurants.map(r => ({ value: r.id, label: r.RestaurantName }))
  );
export const getCategoriesSelect = Restaurant_id =>
  getDishCategoriesByRestId(Restaurant_id).then(categories =>
    categories.map(c => ({ value: c.id, label: c.CategoryName }))
  );
export const getDishesSelect = Category_id =>
  getDishesByDishCategoryId(Category_id).then(dishes =>
    dishes.map(d => ({ value: d.id, label: d.DishName }))
  );
