// api/shopping-cart.js
import * as R from "ramda";
import { cartStore } from "store";

const dishProps = [
  "name",
  "count",
  "sizes",
  "size_id",
  "products",
  "addlProducts",
  "defaultPhoto",
  "RestaurantName",
  "RestaurantDish_id"
];
const emptyCart = {
  dishes: [],
  totalPrice: 0
};

export const setCart = value => {
  const cartWithPromocode =
    cartStore.cart && R.has("promoCode", cartStore.cart)
      ? { ...value, promoCode: cartStore.cart.promoCode }
      : undefined;

  const cartValue = cartWithPromocode ? setDiscount(cartWithPromocode) : value;
  localStorage.setItem("cart", JSON.stringify(cartValue));
  const cart = JSON.parse(localStorage.getItem("cart"));

  cartStore.updateCart(cart);
};

export const setPromoCode = promoCode => {
  R.compose(
    setCart,
    setDiscount,
    R.assoc("promoCode", promoCode)
  )(cartStore.cart);
};
const setDiscount = cart => {
  let extract = 0;
  cart.promoCode.dishes_in_menus.forEach(({ dishes_in_promotion: promo }) => {
    if (R.contains(promo.MenuDish_id, R.map(R.prop("id"), cart.dishes)))
      extract +=
        (R.find(R.propEq("RestaurantDish_id", promo.MenuDish_id), cart.dishes).totalPrice *
          promo.DiscountValue) /
        100;
  });
  return R.assoc("promoPrice", cart.totalPrice - extract, cart);
};

const createNewCart = () => {
  setCart(emptyCart);
  return emptyCart;
};

const getSize = d => R.find(size => d.size_id == size.id)(d.sizes);

const countPrice = d => getSize(d).price * d.count;

const getDish = (d_id, dishes) => R.find(R.propEq("id", d_id))(dishes);

export const addDishToCart = propDish => {
  const cCart = cartStore.cart;
  const cart = cCart ? cCart : createNewCart();
  // const propDish = R.pick(dishProps, dish);
  const dishSize = getSize(propDish);
  const newDishPrice = countPrice(propDish);
  const oldDish = getDish(dishSize.DishInMenuId, cart.dishes);
  const newDish = !oldDish
    ? {
        ...propDish,
        id: dishSize.DishInMenuId,
        price: dishSize.price,
        totalPrice: newDishPrice,
        size: dishSize.name
      }
    : {
        ...oldDish,
        count: oldDish.count + propDish.count,
        totalPrice: oldDish.totalPrice + newDishPrice
      };
  const newDishes = oldDish
    ? R.map(R.ifElse(R.propEq("id", oldDish.id), () => newDish, item => item))(
        cart.dishes
      )
    : [...cart.dishes, newDish];
  setCart({ dishes: newDishes, totalPrice: cart.totalPrice + newDishPrice });
};
export const udpateDishInCart = (oldDish, newDish) => {
  dishDelete(oldDish.id);
  addDishToCart(newDish);
};
const dishUpDown = (id, dec) => {
  const oldCart = cartStore.cart;
  let totalPrice = oldCart.totalPrice;
  const addSub = !dec ? R.add : R.subtract;
  const newDishes = R.map(
    R.ifElse(
      R.propEq("id", id),
      dish => {
        if (dish.count === 1 && dec) return dish;
        totalPrice = addSub(totalPrice, dish.price);
        return {
          ...dish,
          count: addSub(dish.count, 1),
          totalPrice: addSub(dish.totalPrice, dish.price)
        };
      },
      item => item
    )
  )(oldCart.dishes);
  setCart({ dishes: newDishes, totalPrice });
};

export const dishUp = id => dishUpDown(id, false);
export const dishDown = id => dishUpDown(id, true);
export const dishDelete = id => {
  const oldCart = cartStore.cart;
  const dish = getDish(id, oldCart.dishes);
  setCart({
    dishes: R.reject(R.eqProps("id", dish), oldCart.dishes),
    totalPrice: oldCart.totalPrice - dish.totalPrice
  });
};
