// shared-components/products-list_old/schedule-item-products-list.js
import React from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemSecondaryAction from "@material-ui/core/ListItemSecondaryAction";
import ListItemText from "@material-ui/core/ListItemText";
import Checkbox from "@material-ui/core/Checkbox";

const styles = theme => ({
  list: {
    width: "100%",
    maxWidth: 360,
    maxHeight: 150,
    overflow: "auto",
    backgroundColor: theme.palette.background.paper
  },
  margin: {
    margin: theme.spacing.unit
  }
});

class CheckboxListSecondary extends React.Component {
  state = {
    checked: []
  };
  getCheckedItems = items => items.filter(item => item.checked);

  componentDidMount(nextProps) {
    this.setState({ checked: this.getCheckedItems(this.props.items) });
  }

  handleToggle = value => () => {
    const { checked } = this.state;
    const currentIndex = checked.indexOf(value);
    const newChecked = [...checked];

    if (currentIndex === -1) {
      newChecked.push(value);
    } else {
      newChecked.splice(currentIndex, 1);
    }

    this.setState({
      checked: newChecked
    });
    this.props.onProductChange(value);
  };

  render() {
    const { classes, items } = this.props;
    const itemsToShow = items ? items : [];
    return (
      <div className={classes.list}>
        <List dense>
          {itemsToShow.map(({ label: name, value: id, checked }) => (
            <ListItem key={id} button>
              <ListItemText primary={name} />
              <ListItemSecondaryAction>
                <Checkbox
                  onChange={() => this.props.onProductChange(id)}
                  checked={!!checked}
                />
              </ListItemSecondaryAction>
            </ListItem>
          ))}
        </List>
      </div>
    );
  }
}

CheckboxListSecondary.propTypes = {
  classes: PropTypes.object.isRequired
};

export default withStyles(styles)(CheckboxListSecondary);
